xonsh/pyghooks.py

```diff
@@ -1628,10 +1630,21 @@ def subproc_arg_callback(_, match):
     yieldVal = Text
     try:
         path = os.path.expanduser(text)
-        path_stat = os.lstat(path)  # lstat() will raise FNF if not a real file
-        yieldVal, _ = color_file(path, path_stat)
+
+        color_file_with_lstat = lambda p: color_file(
+            p, os.lstat(p)
+        )  # lstat() will raise FNF if not a real file
+
+        pool = ThreadPool(processes=1)
+        color_async = pool.apply_async(color_file_with_lstat, (path,))
```

Comment:
this will slowdown in other use-cases right?

Reply:
@jnoortheen Yes it will.

I wrote a script to benchmark the effects of using `ThreadPool`: [benchmark_threading_lstat.py](https://gist.github.com/yaxollum/3edbbb3e41f70f7996bba800e019cd1c)

The script tests three different methods:
1. Creating a new `ThreadPool` every time the function is called (this is what my PR currently does)
2. Creating only a single `ThreadPool` and using it for every function call
3. Not using any threads at all

Here are the results which I got (time per `lstat` call):
```
With thread new pool:    0.6194 ms
With thread single pool: 0.0770 ms
Without thread:          0.0013 ms
```
As you can see, using a single `ThreadPool` improves the performance by around 8x compared to creating a new `ThreadPool` each time, but that is still around 60x slower than not using any threads.

I'm not sure whether this slowdown is acceptable since although there is a big performance difference ratio, the magnitudes are all under a millisecond.

Reply:
Also, I just realized that using a single `ThreadPool` wouldn't work in the case of `lstat` calls. When an `lstat` call times out, the thread is still running (it doesn't get killed), which would keep the `ThreadPool` occupied. Subsequent calls to valid paths would be queued and also time out.

Reply:
you can hold reference to the previous future and cancel if it was pending after new request comes in.
Also making this optional will help users who work with slow fs.

Reply:
> you can hold reference to the previous future and cancel if it was pending after new request comes in.

What if the previous future is already executing? As far as I know, [futures that have already started executing cannot be canceled](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.cancel):

_"Attempt to cancel the call. **If the call is currently being executed or finished running and cannot be cancelled** then the method will return False, otherwise the call will be cancelled and the method will return True."_

> Also making this optional will help users who work with slow fs.

Do you mean making an environment variable like `$THREAD_COLORING_STAT_CALLS`? Should it be enabled or disabled by default?

Reply:
> What if the previous future is already executing?

can use [killable threads](http://tomerfiliba.com/recipes/Thread2/). But I need opinion on this from others as well. We can also give a warning that stat calls are slow and can suggest to turn-off coloring.

> Should it be enabled or disabled by default?
I suggest to make the default debhaviour to not thread any call.

Reply:
> can use killable threads.

Could another option be to just let the thread continue to run in the background (without killing it)?

Reply:
That will exhaust resources if stuck.